src/support/allocators/pool.h

```diff
@@ -164,6 +174,21 @@ class PoolResource final
         m_allocated_chunks.emplace_back(m_available_memory_it);
     }
 
+    // copied from src/memusage.h
+    size_t MallocUsage(size_t alloc) const
+    {
+        // Measured on libc6 2.19 on Linux.
```

Comment:
Not sure what value this comment adds. Also don't see a reason to mention a version of glibc that we don't support?

Reply:
I copied this function from `memusage.h` ([here](https://github.com/bitcoin/bitcoin/blob/6cf47a8f44f96099a84e5c6e628e3499045e024d/src/memusage.h#L52)). I'm hoping not to duplicate this function, even though it's pretty small. `pool.h` can't include `memusage.h` because that creates a circular dependency. I'll try to figure out a way to improve this and also of course I welcome suggestions.

Anyway, if we need to duplicate this function, I'll remove this comment, I agree it's not useful.

Reply:
About that function, I do not think that it is accurate any more. I wrote a different version of that estimation that makes use of `malloc_usable_size`, and that way I get different values, see here: https://godbolt.org/z/3a6E57dPx

It might be worth having a look if using that estimation is more accurate

Reply:
Thanks, I got a slightly different result -- `malloc_usable_size() + sizeof(void*)` is perfect (note adding only `sizeof(void*)` instead of 2 times that). I'm not set up for 32-bit, can you try running this test program on 32-bit if you are able to easily?

Whichever is the most accurate, would a PR to improve [`memusage.h: MallocUsage()`](https://github.com/bitcoin/bitcoin/blob/6cf47a8f44f96099a84e5c6e628e3499045e024d/src/memusage.h#L50) be worth doing? (I would say a separate PR would be better than doing that here; then this PR could be based on that one.)

<details>
<summary>malloc space usage estimation test program</summary>

```
// malloc space usage estimation
//
// g++ -std=c++17 -m64 t.cpp
//
// See:
// https://github.com/bitcoin/bitcoin/pull/27748#discussion_r1209810958
//
// This program does a bunch of memory allocations of various
// sizes, and calculates the memory usage increase using /proc.
//
// For each allocation size, it prints:
//   allocation size
//   malloc_usable_size() + sizeof(void*)
//   empirical estimate
//   empirical actual
//
// This seems to show that, at least for x86_64, an accurate estimate is:
//
//   malloc_usable_size() + sizeof(void*)
//
// I don't know why we need to add sizeof one void* instead of two.
// I didn't try this with -m32 (32-bit model).
// (LarryRuane@gmail.com)
//
#include <malloc.h>
#include <algorithm>
#include <iostream>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <unistd.h>

size_t heapsize()
{
    std::ifstream maps("/proc/" + std::to_string(getpid()) + "/maps");
    std::string line;
    while (std::getline(maps, line)) {
        if (line.find("[heap]") == std::string::npos) continue;
        // example:
        // 563579244000-56358c384000 rw-p 00000000 00:00 0 [heap]
        std::string::size_type dash_pos{line.find('-')};
        std::string::size_type space_pos{line.find(' ')};
        std::string start_str{line.substr(0, dash_pos)};
        std::string end_str{line.substr(dash_pos+1, space_pos-dash_pos-1)};

        size_t start{std::stoul(start_str, nullptr, 16)};
        size_t end{std::stoul(end_str, nullptr, 16)};
        return end - start;
    }
    return 0;
}

constexpr size_t n_allocations{100'000};
int main(int argc, char** argv) {
    for (size_t s{0}; s < 120; ++s) {
        size_t start{heapsize()};
        void* p;
        for (int i = 0; i < n_allocations; ++i)
            p = malloc(s);
        size_t end{heapsize()};
        double actual{double(end-start) / n_allocations};

        // empirically-derived, 64-bit linux x86_64
        size_t estimate{std::max<size_t>(32, ((s+7)/16+1)*16)};
        std::cout << s << ' '
            << malloc_usable_size(p) + sizeof(void*) << ' '
            << estimate << ' '
            << actual << '\n';
    }
    return 0;
}
```
</details>